pkg/cloud/generic/actuators/machine/actuator.go

```diff
@@ -52,32 +52,72 @@ func NewActuator(params ActuatorParams) (*Actuator, error) {
 
 // Create creates a machine and is invoked by the Machine Controller
 func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+	// Create is only called if a Machine resource exists. It is a
+	// configuration error if a Cluster does not exist in the same
+	// namespace as the Machine. By returning an error here we ensure
+	// the resource will be requeued for processing later.
+	if cluster == nil {
+		return fmt.Errorf("machine %s does not have cluster in namespace %s", machine.Name, machine.Namespace)
+	}
+
 	log.Printf("Creating machine %v for cluster %v.", machine.Name, cluster.Name)
-	return fmt.Errorf("TODO: Not yet implemented")
+
+	providerID := fmt.Sprintf("%s-%s", cluster.Name, machine.Name)
+	machine.Spec.ProviderID = &providerID
+	machine, err := a.machinesGetter.Machines(machine.Namespace).Update(machine)
+	if err != nil {
+		log.Printf("Failed to set Machine.Spec.ProviderID %s: %v", providerID, err)
+		return fmt.Errorf("failed to set Machine.Spec.ProviderID %s: %v", providerID, err)
+	}
+
+	// TODO: Call webhook to allocate infrastruction
+
+	return nil
 }
 
 // Delete deletes a machine and is invoked by the Machine Controller
 func (a *Actuator) Delete(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+	if cluster == nil {
+		return fmt.Errorf("machine %s does not have cluster in namespace %s", machine.Name, machine.Namespace)
+	}
+
 	log.Printf("Deleting machine %v for cluster %v.", machine.Name, cluster.Name)
-	return fmt.Errorf("TODO: Not yet implemented")
+
+	if machine.Spec.ProviderID == nil {
```

Comment:
Why is this done?

The normal sequence of events:
1. Machine controller calls `Delete(ctx, c, m)`
2. Actuator calls webhook to delete the infra for `m`
3. Actuator removes `ProviderID` from `m`.
4. Machine controller removes the finalizer from `m`.
5. The API server deletes the Machine object for `m`.

If (3) is not done, then I suppose the sequence might start over, if, for example, the Machine controller crashes immediately before (4). But the webhook delete is idempotent, so that's ok.

Reply:
I might not understand.`Create()` sets `ProviderID` before calling the webhook. So if `ProviderID` is not set then there is nothing to clean up. In addition to the controller crashing, it is possible for `Delete()` to be called before `Create()` since events are level triggered.